### PR TITLE
[JENKINS-55348] Use supplied object display names for the job

### PR DESCRIPTION
**Pull request by cobexer**
If the SCMHead has an ObjectMetadataAction that defines a display name use it for the job display name

This causes jobs for GitLab Merge Requests, GitHub Pull Requests, ... to display like:
"MR-1234 - Enhanced Job Display Names"
inspired by: https://github.com/jenkinsci/gitlab-branch-source-plugin/issues/141

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by jglick — COMMENTED** [review]
As mentioned on the dev list, this could have unwanted effects, such as the display name being too long to be displayed comfortably in a variety of contexts such as the executor widget. Probably too crude a tool. Suggest setting `Job.description`, or adjust https://github.com/jenkinsci/branch-api-plugin/blob/1b883b9a05b7c1a265975eb30e1da9eaa0fa3f03/src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java#L222-L224 to include a new view column showing `ObjectMetadataAction.objectDisplayName`.

**Comment by jglick** [comment]
Disclaimer: I am **not** a maintainer of this plugin.

**Comment by cobexer** [comment]
> As mentioned on the dev list, this could have unwanted effects, such as the display name being too long to be displayed comfortably in a variety of contexts such as the executor widget. Probably too crude a tool. Suggest setting `Job.description`, or adjust
> 
> https://github.com/jenkinsci/branch-api-plugin/blob/1b883b9a05b7c1a265975eb30e1da9eaa0fa3f03/src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java#L222-L224
> to include a new view column showing `ObjectMetadataAction.objectDisplayName`.

I have considered your suggestion but decided against it because:
 * The Jobs display name is reliably shown in all places where the job is referenced (we consider this essential)
 * our jenkins users always have to include issue identifiers in all branch names and commit messages which enables everyone to correlate changes to the relevant Story / Defect in our issue tracker
 * Job names are already way too long and the display issues are already happening so this is a minor worsening for us
 * Feature branches have build names that also include the branch name ... so every run everywhere has a ridiculously long name already 😅
 * Jenkins BUILD_NR or GitLab's MR numbers have no meaning to us at all

I do however understand that this may be too much, which is why I pushed this change this early to gather feedback 🙂

@bitwiseman You seem to be the active maintainer of this plugin: What's your take on this?

Test errors, code duplication, missing tests and missing related cleanup: come after a general ok for the selected approach 😉

**Comment by cobexer** [comment]
@bitwiseman I'm still waiting for an ok for the general approach! Please have a look and let me know =)

**Comment by bitwiseman** [comment]
@cobexer 
I think this is interesting and potentially useful behavior and having it based in this plugin and then supplied by branch source plugins is better than having each branch source go it alone.  So overall, good choice.  

This is certainly better than the alternative I've seen that actually changes job name, rather than just the display name. 

Still, I think this behavior should be controlled/configurable by a setting/trait.  That way people who believe the behavior is useful can enable it, but it will not bother or break those who do not want it.

**Comment by jglick** [comment]
> controlled/configurable by a setting/trait

IIRC you can create an `SCMTrait` applicable to both repo folders and org folders.

**Comment by cobexer** [comment]
> @cobexer
> Could you provide some screenshots show the good and the bad of the new display names?

@bitwiseman I did some digging and attached the screenshots to the existing issue: https://issues.jenkins.io/browse/JENKINS-55348

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 1]
```diff
@@ -1999,6 +1999,21 @@ private void observeExisting(@NonNull SCMHead head, @NonNull SCMRevision revisio
...
                     .equals(Util.getDigestOf(Items.XSTREAM2.toXML(origBranch.getScm())));
             _factory.decorate(_factory.setBranch(project, branch));
+
+            String displayName = null;
+            try {
+                ObjectMetadataAction action = project.getAction(ObjectMetadataAction.class);
+                if (action != null) {
+                    displayName = action.getObjectDisplayName();
+                    if (!StringUtils.isBlank(displayName)) {
+                        displayName = rawName + " - " + displayName;
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java:2009]:
```suggestion
                        displayName = rawName + ": " + displayName;
```

or
```suggestion
                        displayName = rawName + "—" + displayName;
```

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 2]
```diff
@@ -2068,6 +2083,13 @@ private void observeNew(@NonNull SCMHead head, @NonNull SCMRevision revision, @N
                 if (project.getDisplayNameOrNull() == null && !rawName.equals(encodedName)) {
                     project.setDisplayName(rawName);
                 }
+                ObjectMetadataAction action = project.getAction(ObjectMetadataAction.class);
+                if (action != null) {
+                    String displayName = action.getObjectDisplayName();
+                    if (!StringUtils.isBlank(displayName)) {
+                        project.setDisplayName(rawName + " - " + displayName);
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java:2090]:
```suggestion
                        project.setDisplayName(rawName + ": " + displayName);
```

or
```suggestion
                        project.setDisplayName(rawName + "—" + displayName);
```

**Comment by cobexer** [comment]
@bitwiseman @jglick I made the Job display name behaviour conditional on the option seleted with a trait, keeping the default behaviour unchanged. Next week we will deploy this version to our company Jenkins to gather more internal feedback.
Please take a moment to review the current version

**Comment by per-bohlin** [comment]
When can we expect to see this feature released?

**Comment by cobexer** [comment]
BTW: This version is running in our Jenkins instance since my last comment without issues :)

@rsandell could you do the review maybe?

**Comment by quilicicf** [comment]
I'd be interested in having this feature, too :pray: 
It looks safe since it's been tested and the default behavior doesn't change. 
Are there plans to release this?

**Comment by quilicicf** [comment]
@bitwiseman Any way to get this merged? :pray:

**Comment by acatoire** [comment]
This would help a lot, the actual GitHub PR view is really not efficient on big projects.
image

**Comment by quilicicf** [comment]
@jglick Do you know what it'd take to get this merged? 
I can and am willing to spend some time to push this forward but I don't know what's missing, tests?
I had a look at the tests and it looks like there are currently no tests with pull requests.
From what I understand, writing one would require importing a plugin that supports pull requests like the `github-branch-source` plugin?

My understanding of how one should write those tests is currently:
1. Add a plugin with pull request support like `github-branch-source` with scope test
2. Create a test class with a structure like [OverrideIndexTriggersJobPropertyTest](https://github.com/jenkinsci/branch-api-plugin/blob/db1bc433ea76d05c84ab4c6aa5351566dfe45147/src/test/java/jenkins/branch/OverrideIndexTriggersJobPropertyTest.java#L189) but two with two repos and a fake PR from one repo to the other
3. Test that the PR that's generated has the right name

Since there are no tests with PRs in the repo currently, I can't copy-paste much so should I attempt to implement this test, I'd need some guidance.

Am I on the right track? Would you like me to try to implement this test?

**Comment by jglick** [comment]
https://github.com/jenkinsci/branch-api-plugin/pull/254#issuecomment-852523365

**Comment by quilicicf** [comment]
:disappointed: Saw you merge PRs on this repo, thought you could help. 
@bitwiseman @rsandell Could you have a look at my previous message? Can any of you help me move this forward? :pray:

**Comment by jglick** [comment]
> Saw you merge PRs on this repo

Yes but only trivial metadata changes needed to unblock test tooling and the like, not substantive PRs.

**Comment by quilicicf** [comment]
I opened [a new pull request](https://github.com/jenkinsci/branch-api-plugin/pull/374) that builds upon this one and adds a test. 

Hopefully, this makes the feature request complete and can be merged :crossed_fingers:

**Comment by jtnord** [comment]
obsolete (and included) by https://github.com/jenkinsci/branch-api-plugin/pull/374

**Pull request closed without merging**